### ISSUE TITLE
Unable to transform test data after MCA fitting the training data

### ISSUE THREAD

**Issue by adeebabdulsalam** [MaxHalford/prince#53]
```
data = pd.read_csv("data/training set.csv")
X = data.loc[:, 'OS.1':'DSA.1']

df = pd.DataFrame(X)

mca = prince.MCA(
               n_components=2,
               n_iter=3,
               copy=True,
               check_input=True,
               engine='auto',
               random_state=42
                )

mca = mca.fit(df)

df_new = df.loc[0:5, :]
I = mca.transform(df_new)
print(I)

```

Output: 
File "C:/../clustering/k means.py", line 62, in <module>
    I = mca.transform(df_new)
  File "C:\..\clustering\interpreter2\lib\site-packages\prince\mca.py", line 47, in transform
    return self.row_coordinates(X)
  File "C:\..\clustering\interpreter2\lib\site-packages\prince\mca.py", line 37, in row_coordinates
    return super().row_coordinates(self.one_hot_.transform(X))
  File "C:\..\clustering\interpreter2\lib\site-packages\prince\ca.py", line 111, in row_coordinates
    X = X / X.sum(axis=1)
  File "C:\python36\lib\site-packages\scipy\sparse\base.py", line 1015, in sum
    np.ones((n, 1), dtype=res_dtype))
  File "C:\python36\lib\site-packages\scipy\sparse\base.py", line 499, in __mul__
    result = self._mul_vector(np.ravel(other))
  File "C:\python36\lib\site-packages\scipy\sparse\coo.py", line 571, in _mul_vector
    other.dtype.char))
  File "C:\python36\lib\site-packages\scipy\sparse\sputils.py", line 60, in upcast_char
    t = upcast(*map(np.dtype, args))
  File "C:\python36\lib\site-packages\scipy\sparse\sputils.py", line 52, in upcast
    raise TypeError('no supported conversion for types: %r' % (args,))
TypeError: no supported conversion for types: (dtype('O'), dtype('O'))

This is how data looks like
```
print(df_new)

output:
    0  1   2   3
0  9  8   9   9
1  8  7   8   6
2  8  7   9   9
3  8  7   9   9
4  8  7   8   7
5  9  8  10  10

```

python 3.6.4
scikit 0.20.2
numpy 1.16.1
pandas  0.24.1

**Comment by MaxHalford** [comment]
From a first look it looks like `df_new` doesn't have the same structure as `df`. What exactly are you truing to do?

**Comment by adeebabdulsalam** [comment]
> From a first look it looks like `df_new` doesn't have the same structure as `df`. What exactly are you trying to do?

df is GRADES(S=10,A=9,B=8 ... etc) of students in 4 different subjects.
df contains 163 students so => 163 rows.

df_new has GRADES of 4 subjects of 1 or more students.
I want to reduce dimensions of df_new and predicts it's cluster in KMEANS

**Comment by adeebabdulsalam** [comment]
> From a first look it looks like `df_new` doesn't have the same structure as `df`. What exactly are you trying to do?

Ok, so i made one change to df_new .Took initial 6 rows of df itself as df_new.
This time the error is different

**Comment by adeebabdulsalam** [comment]
Hey ,
Thank you so much for your time and effort.
Really appreciate it.
But I'm having issues with it. And my final year project depends on it
.Since you are the one who did it, there is no other way for me to find a
solution.please help when it is possible
Giving the link to the issue below

https://github.com/MaxHalford/prince/issues/53#issuecomment-465108819

**Comment by MaxHalford** [comment]
Can you provide me with the CSV file? Also please stop editing your answers and write new ones, it makes your issue hard to follow otherwise.

**Comment by adeebabdulsalam** [comment]
> Can you provide me with the CSV file? Also please stop editing your answers and write new ones, it makes your issue hard to follow otherwise.

https://drive.google.com/open?id=1GWNax2NI84g0tfHXNKQKFPAcYDu9zN_fp1QawskAq30

Here is the doc to the csv i used.But i have preproccesed the grades into numbers in python like i said before.

And sorry about the editing,first time using the Github Issues

**Comment by kuhanw** [comment]
Hi, 

I can confirm I am having this exact problem. Even when I try to use the tutorial example of 

X = pd.DataFrame(
     data=[
         [1, 6, 7, 2, 5, 7, 6, 3, 6, 7],
         [5, 3, 2, 4, 4, 4, 2, 4, 4, 3],
         [6, 1, 1, 5, 2, 1, 1, 7, 1, 1],
         [7, 1, 2, 7, 2, 1, 2, 2, 2, 2],
         [2, 5, 4, 3, 5, 6, 5, 2, 6, 6],
         [3, 4, 4, 3, 5, 4, 5, 1, 7, 5]
     ],
     columns=['E1 fruity', 'E1 woody', 'E1 coffee',
              'E2 red fruit', 'E2 roasted', 'E2 vanillin', 'E2 woody',
              'E3 fruity', 'E3 butter', 'E3 woody'],
     index=['Wine {}'.format(i+1) for i in range(6)]
 )

If I fit on the entire dataset:

mca.fit(X)

then try to transform only two specific rows:


---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-162-cf90d090c615> in <module>()
----> 1 mca.transform(X[5:7])

c:\users\wangku\appdata\local\continuum\anaconda3\envs\tensorflow-gpu\lib\site-packages\prince\mca.py in transform(self, X)
     45         if self.check_input:
     46             utils.check_array(X, dtype=[str, np.number])
---> 47         return self.row_coordinates(X)
     48 
     49     def plot_coordinates(self, X, ax=None, figsize=(6, 6), x_component=0, y_component=1,

c:\users\wangku\appdata\local\continuum\anaconda3\envs\tensorflow-gpu\lib\site-packages\prince\mca.py in row_coordinates(self, X)
     35 
     36     def row_coordinates(self, X):
---> 37         return super().row_coordinates(self.one_hot_.transform(X))
     38 
     39     def column_coordinates(self, X):

c:\users\wangku\appdata\local\continuum\anaconda3\envs\tensorflow-gpu\lib\site-packages\prince\ca.py in row_coordinates(self, X)
    109             X = X / X.sum(axis=1)[:, None]
    110         else:
--> 111             X = X / X.sum(axis=1)
    112 
    113         return pd.DataFrame(

c:\users\wangku\appdata\local\continuum\anaconda3\envs\tensorflow-gpu\lib\site-packages\scipy\sparse\base.py in sum(self, axis, dtype, out)
   1013             # sum over rows
   1014             ret = self * np.asmatrix(
-> 1015                 np.ones((n, 1), dtype=res_dtype))
   1016 
   1017         if out is not None and out.shape != ret.shape:

c:\users\wangku\appdata\local\continuum\anaconda3\envs\tensorflow-gpu\lib\site-packages\scipy\sparse\base.py in __mul__(self, other)
    497                 raise ValueError('dimension mismatch')
    498 
--> 499             result = self._mul_vector(np.ravel(other))
    500 
    501             if isinstance(other, np.matrix):

c:\users\wangku\appdata\local\continuum\anaconda3\envs\tensorflow-gpu\lib\site-packages\scipy\sparse\coo.py in _mul_vector(self, other)
    569         #output array
    570         result = np.zeros(self.shape[0], dtype=upcast_char(self.dtype.char,
--> 571                                                             other.dtype.char))
    572         coo_matvec(self.nnz, self.row, self.col, self.data, other, result)
    573         return result

c:\users\wangku\appdata\local\continuum\anaconda3\envs\tensorflow-gpu\lib\site-packages\scipy\sparse\sputils.py in upcast_char(*args)
     58     if t is not None:
     59         return t
---> 60     t = upcast(*map(np.dtype, args))
     61     _upcast_memo[args] = t
     62     return t

c:\users\wangku\appdata\local\continuum\anaconda3\envs\tensorflow-gpu\lib\site-packages\scipy\sparse\sputils.py in upcast(*args)
     50             return t
     51 
---> 52     raise TypeError('no supported conversion for types: %r' % (args,))
     53 
     54 

TypeError: no supported conversion for types: (dtype('O'), dtype('O'))

I think this might have something to do with scipy versions? Googling, there seems to be similar errors for other use cases. Could you tell us what versions of the dependency packages you used? (Numpy, scipy, etc...)

**Comment by MaxHalford** [comment]
Can you try with the latest version? I think this should be fixed now.

**Comment by kuhanw** [comment]
Thanks for the fix. This seems to work for MCA transform now. FAMD transform still has the issue from:https://github.com/MaxHalford/prince/issues/56.

**Comment by MaxHalford** [comment]
@adeebabdulsalam can you see if the latest version fixes your issue (if you have time of course)?

**Comment by adeebabdulsalam** [comment]
@MaxHalford   Thank you so much!!!! It's working

**Comment by MaxHalford** [comment]
Great! Sorry this took so much time. Have a very nice day!
